feat(editor-3001): sidebar folders

EDsCODE opened the pull request:
## Problem

- can't organize queries when I have a ton of queries

<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

- create folders to organize

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

EDsCODE committed:
folder wip

EDsCODE committed:
working folders

EDsCODE committed:
sidebar using folders

EDsCODE committed:
typing

EDsCODE committed:
typing

EDsCODE committed:
Merge branch 'master' into editor-3001-folders

EDsCODE committed:
change naming

EDsCODE committed:
Merge branch 'editor-3001-folders' of github.com:PostHog/posthog into editor-3001-folders

EDsCODE committed:
make search alwas visible

EDsCODE committed:
add padding

EDsCODE committed:
make tabs look like tabs

EDsCODE committed:
merge base

EDsCODE committed:
resort the sidebar contents. add posthog under 'sources'

EDsCODE committed:
wip

EDsCODE committed:
wip

EDsCODE committed:
wip

EDsCODE committed:
merge main

github-actions[bot] committed:
Update query snapshots

EDsCODE committed:
adjust api for test

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

EDsCODE committed:
order alphabetically

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

EDsCODE committed:
working state

EDsCODE committed:
stateful folders

EDsCODE committed:
merge main

EDsCODE committed:
reset pnpm lock

EDsCODE committed:
typing

EDsCODE committed:
Merge branch 'master' into editor-3001-create-folders

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

phixMe commented on frontend/src/scenes/data-warehouse/editor/editorSidebarLogic.tsx:
```diff
@@ -62,139 +57,273 @@ export const editorSidebarLogic = kea<editorSidebarLogicType>([
+                    const folder = await api.dataWarehouseFolders.create({ name, parent: parentId })
+                    return values.folders.concat(folder)
+                },
+                deleteFolder: async ({ id }) => {
+                    await api.dataWarehouseFolders.delete(id)
+                    return values.folders.filter((folder) => folder.id !== id)
+                },
```
I'm able to generate stacktraces when removing nested empty folders:
<img width="514" alt="image" src="https://github.com/user-attachments/assets/b5f7f27c-44ba-4cb4-96b9-1d04c39eb252" />

phixMe commented on frontend/src/scenes/data-warehouse/editor/editorSidebarLogic.tsx:
```diff
@@ -62,139 +57,273 @@ export const editorSidebarLogic = kea<editorSidebarLogicType>([
+                    const response = await api.dataWarehouseFolders.list()
+                    return response.results
+                },
+                addFolder: async ({ name, parentId }) => {
+                    const folder = await api.dataWarehouseFolders.create({ name, parent: parentId })
+                    return values.folders.concat(folder)
```
I also attempted to add multiple new folders at the same time and ran into the duplicate name issue which I think could be resolved by some server side intelligence by returning something like `New folder (1)` like an operating system would.

EDsCODE replied:
changed workflow to be require you to provide a name immediately like in a code editor

phixMe commented on frontend/src/layout/navigation-3000/components/SidebarList.tsx:
```diff
@@ -537,23 +614,81 @@ function SidebarListItemSkeleton({ style }: { style: React.CSSProperties }): JSX
+                }
+            }
+        }
+    }, [isRenaming, category.onRename, startRenaming])
```
There is some propagation behavior on this where clicking once toggles the openness of the folder and the double click is what is needed for editing. I ran into some of this with the tab renaming, maybe it would best to discuss a system to handle this kind of thing generally, with a consistent edit state that isn't too disruptive.

EDsCODE replied:
yeah agreed. will roll with this for now but will need to revisit the interaction behaviors

phixMe commented on frontend/src/scenes/data-warehouse/editor/editorSidebarLogic.tsx:
```diff
@@ -62,139 +57,273 @@ export const editorSidebarLogic = kea<editorSidebarLogicType>([
+                                        name: savedQuery.name,
+                                        url: '',
+                                        icon: savedQuery.status ? (
+                                            <Tooltip title="Materialized view">
+                                                <IconCalculate />
+                                            </Tooltip>
```
I'm having console errors flushed and I think they are related to this one?
<img width="517" alt="image" src="https://github.com/user-attachments/assets/be0eb9b2-897b-477c-8591-00520cb5cacf" />

phixMe reviewed (commented):
This is neat! Nice work!

EDsCODE committed:
DRY

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

EDsCODE committed:
merge master

EDsCODE committed:
add validation

EDsCODE committed:
fix deletion

EDsCODE committed:
add temp files

EDsCODE commented on frontend/src/layout/navigation-3000/components/SidebarList.tsx:
```diff
@@ -50,6 +55,62 @@ export function SidebarList({ category }: { category: SidebarCategory | ListItem
+
+        // Get the last item in the chain of item separator
+        const getLastItemInChain = (key: string): string => {
+            const parts = key.split(ITEM_KEY_PART_SEPARATOR)
```
added validation to restrict folder naming schemes

EDsCODE commented on frontend/src/layout/navigation-3000/components/SidebarList.tsx:
```diff
@@ -50,6 +55,62 @@ export function SidebarList({ category }: { category: SidebarCategory | ListItem
+        setActiveId(event.active.id.toString())
+    }
 
     const emptyStateSkeletonCount = useMemo(() => 4 + Math.floor(Math.random() * 4), [])
```
Just had some numbers generating to fill space. keeping it at 4 now

EDsCODE committed:
add back actions

EDsCODE committed:
fix div issues

EDsCODE commented on frontend/src/layout/navigation-3000/components/SidebarList.tsx:
```diff
@@ -50,6 +55,62 @@ export function SidebarList({ category }: { category: SidebarCategory | ListItem
+        const { active, over } = event
+
+        // Get the last item in the chain of item separator
+        const getLastItemInChain = (key: string): string => {
+            const parts = key.split(ITEM_KEY_PART_SEPARATOR)
+            return parts[parts.length - 1]
+        }
```
hmm not following here. I use this function in both parts of the if/else

phixMe replied:
You're right!

EDsCODE committed:
adjustments

EDsCODE committed:
Merge branch 'master' into editor-3001-create-folders

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

EDsCODE committed:
fix types

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

EDsCODE committed:
strong delete

EDsCODE committed:
add confirmation

EDsCODE committed:
more type issues

phixMe commented:
I'm experiencing some issues with the `AuthSizer` which may be out of scope here where opening the view folder covers/obscures the sources entirely.

### Open
<img width="301" alt="image" src="https://github.com/user-attachments/assets/61ab3cd1-8c8f-45f8-97c6-568acf6ee870" />

### Closed
<img width="303" alt="image" src="https://github.com/user-attachments/assets/c5ea885d-ab30-4a38-9b76-20bf657f5644" />

I also seem to be seeing some inaccuracy of the measurement in the AutoSizer and some of the bottom labels are not clickable.
<img width="305" alt="image" src="https://github.com/user-attachments/assets/105c63ac-30be-4954-867d-1ea2386246b6" />

Also, for the sources they seem to be `grabbable` in the same way as the views, but they aren't re-categorizable in the way that views are. I think there a number of ways to depict what is and isn't allowed like disabling the grab, or adding a handle icon to the view items on the left hand side.

EDsCODE commented:
> I'm experiencing some issues with the AuthSizer which may be out of scope here where opening the view folder covers/obscures the sources entirely.

When views are open, do you see the sources at the bottom?

EDsCODE committed:
add draggable

EDsCODE commented:
Fixed the above! Added a draggable handle on hover and made it so that you can only drag items that are meant to be dragged

![CleanShot 2025-01-23 at 14 35 34@2x](https://github.com/user-attachments/assets/68355c3e-2f6f-46a5-adbb-31668d4118a1)

EDsCODE committed:
100% instead of 100vh

EDsCODE committed:
hide overflow

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

phixMe reviewed (approved):
Thanks for all the effort here. Works and looks great!

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

EDsCODE committed:
update fields

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

EDsCODE committed:
merge main

EDsCODE committed:
update

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

EDsCODE committed:
Merge branch 'master' into editor-3001-create-folders

EDsCODE committed:
merge main

EDsCODE committed:
Merge branch 'editor-3001-create-folders' of github.com:PostHog/posthog into editor-3001-create-folders

EDsCODE committed:
merge main

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)
